src/Magnum/Platform/EmscriptenApplication.cpp

```diff
@@ -0,0 +1,505 @@
+    {
+        return found->value;
+    }
+
+    /* F1 - F12 */
+    if(Utility::String::beginsWith(code, "F")) {
+        /* F1-F9 */
+        if(code[2] != '\0') {
+            const Int num = code[2] - '0';
+            return Key(Int(Key::F10) + num);
+        }
+        /* F10-F12 */
+        const Int num = code[1] - '1';
+        return Key(Int(Key::F1) + num);
+    }
+
+    return Key::Unknown;
+}
+
+void EmscriptenApplication::startTextInput() {
+    _flags |= Flag::TextInputActive;
+}
+
+void EmscriptenApplication::stopTextInput() {
+    _flags &= ~Flag::TextInputActive;
+}
+
+void EmscriptenApplication::setTextInputRect(const Range2Di& CORRADE_UNUSED(rect)) {
+    // TODO: Place a hidden input field at given rect
+}
+
+void EmscriptenApplication::viewportEvent(ViewportEvent&) {}
+void EmscriptenApplication::keyPressEvent(KeyEvent&) {}
+void EmscriptenApplication::keyReleaseEvent(KeyEvent&) {}
+void EmscriptenApplication::mousePressEvent(MouseEvent&) {}
+void EmscriptenApplication::mouseReleaseEvent(MouseEvent&) {}
+void EmscriptenApplication::mouseMoveEvent(MouseMoveEvent&) {}
+void EmscriptenApplication::mouseScrollEvent(MouseScrollEvent&) {}
+void EmscriptenApplication::textInputEvent(TextInputEvent&) {}
+void EmscriptenApplication::textEditingEvent(TextEditingEvent&) {}
+
+EmscriptenApplication::GLConfiguration::GLConfiguration():
+    _colorBufferSize{8, 8, 8, 0}, _depthBufferSize{24}, _stencilBufferSize{0} {}
+
+void EmscriptenApplication::mainLoopIteration() {
+    /* The resize event is not fired on window resize, so poll for the canvas
+       size here. But only if the window was requested to be resizable, to
+       avoid resizing the canvas when the user doesn't want that. Related
+       issue: https://github.com/kripken/emscripten/issues/1731 */
+    if(_flags & Flag::Resizable) {
+        /* Emscripten 1.38.27 changed to generic CSS selectors from element
+           IDs depending on -s DISABLE_DEPRECATED_FIND_EVENT_TARGET_BEHAVIOR=1
+           being set (which we can't detect at compile time). See above for the
+           reason why we hardcode #canvas here. */
+        const Vector2i canvasSizei{windowSize()};
+        if(canvasSizei != _lastKnownCanvasSize) {
+            _lastKnownCanvasSize = canvasSizei;
+            const Vector2i size = _dpiScaling*canvasSizei;
+            emscripten_set_canvas_element_size("#canvas", size.x(), size.y());
+            ViewportEvent e{size, size, _dpiScaling};
+            viewportEvent(e);
+            _flags |= Flag::Redraw;
+        }
+    }
+
+    if(_flags & Flag::Redraw) {
+        _flags &= ~Flag::Redraw;
+        drawEvent();
+    }
+}
+
+int EmscriptenApplication::exec() {
+    emscripten_set_main_loop_arg([](void* arg) {
+        static_cast<EmscriptenApplication*>(arg)->mainLoopIteration();
+    }, this, 0, true);
+
+    return _exitCode;
+}
+
+void EmscriptenApplication::exit(int exitCode) {
+    emscripten_cancel_main_loop();
+    _exitCode = exitCode;
```

Comment:
Does an exit code make any sense in the context of an in-browser app? I don't think so. The `exitCode` parameter would then be included only for compatibility with other app implementations (and mentioned as such in the docs).